src/App.tsx

```diff
@@ -1,14 +1,73 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useCallback, useEffect, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { FilterBy } from './types/filter';
+
+const filteredTodos = (todos: Todo[], query: string, filterBy: FilterBy) => {
+  let preparedTodos = [...todos];
+
+  if (query) {
+    const preparedQuery = query.trim().toLowerCase();
+
+    preparedTodos = preparedTodos.filter(todo => todo.title.toLowerCase().includes(preparedQuery));
+  }
+
+  switch (filterBy) {
+    case FilterBy.ACTIVE:
+      preparedTodos = preparedTodos.filter(todo => !todo.completed);
+      break;
+    case FilterBy.COMPLETED:
+      preparedTodos = preparedTodos.filter(todo => todo.completed);
+      break;
+    case FilterBy.ALL:
+    default:
+      break;
+  }
+
+  return preparedTodos;
+};
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState('');
+  const [filterBy, setFilterBy] = useState<FilterBy>(FilterBy.ALL);
+  const [isLoadingError, setIsLoadingError] = useState(false);
+  const [isDataLoading, setIsDataLoading] = useState(true);
+
+  const visibleTodos = filteredTodos(todos, query, filterBy);
+
+  const fetchData = async () => {
+    try {
+      const data = await getTodos();
+
+      setTodos(data);
+      setIsDataLoading(false);
```

Comment:
in case with error you always will have isDataLoading = true
<img width="1160" alt="Screenshot 2023-02-27 at 17 19 21" src="https://user-images.githubusercontent.com/53228259/221603824-363a6cde-98e5-420d-9b10-a764f404df0e.png">

better to setIsDataLoading(true) before try and setIsDataLoading(false) in finally